block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala

```diff
@@ -1034,6 +1082,121 @@ object BlockDagFileStorage {
+      for {
+        deployIdSizeOpt <- randomAccessIO.readInt
+        result <- deployIdSizeOpt match {
+                   case Some(deployIdSize) =>
+                     val blockHashBytes = Array.ofDim[Byte](BlockHash.Length)
+                     val deployIdBytes  = Array.ofDim[Byte](deployIdSize)
+                     for {
+                       readDeployId  <- randomAccessIO.readFully(deployIdBytes)
+                       readBlockHash <- randomAccessIO.readFully(blockHashBytes)
+                       result <- (readDeployId, readBlockHash) match {
+                                  case (Some(()), Some(())) =>
+                                    readRec(
+                                      (ByteString.copyFrom(deployIdBytes) -> ByteString
+                                        .copyFrom(blockHashBytes)) :: result
+                                    )
+                                  case _ =>
+                                    Sync[F].raiseError[List[(DeployId, BlockHash)]](
+                                      BlockHashesByDeployLogIsCorrupted
+                                    )
+                                }
+                     } yield result
+                   case None =>
+                     result.reverse.pure[F]
+                 }
+      } yield result
+    readRec(List.empty)
+  }
+
+  def calculateBlockHashesByDeployCrc[F[_]: Monad](
+      blockHashesByDeploy: List[(DeployId, BlockHash)]
+  ): F[Crc32[F]] = {
+    val crc = Crc32.empty[F]()
+    blockHashesByDeploy.traverse_[F, Unit] {
+      case (deployId, blockHash) =>
+        crc.update(deployId.size.toByteString.concat(deployId).concat(blockHash).toByteArray)
+    } >> crc.pure[F]
+  }
+
+  private def truncateBlockHashesByDeployLog[F[_]: Sync](
+      randomAccessIO: RandomAccessIO[F],
+      blockHashesByDeploy: List[(DeployId, BlockHash)]
+  ): F[Unit] = {
+    val lastRecordSize: Long = BlockHash.Length.toLong + BlockHash.Length.toLong
+    for {
+      length <- randomAccessIO.length
+      _      <- randomAccessIO.setLength(length - lastRecordSize)
+    } yield ()
+  }
+
+  private def validateBlockHashesByDeploy[F[_]: Sync](
+      randomAccessIO: RandomAccessIO[F],
+      readBlockHashesByDeployCrc: Long,
+      blockHashesByDeploy: List[(DeployId, BlockHash)]
+  ): F[(List[(DeployId, BlockHash)], Crc32[F])] =
+    for {
+      fullCalculatedCrc <- calculateBlockHashesByDeployCrc[F](blockHashesByDeploy)
+      result <- Monad[F].ifM(fullCalculatedCrc.value.map(_ == readBlockHashesByDeployCrc))(
+                 (blockHashesByDeploy, fullCalculatedCrc).pure[F],
+                 blockHashesByDeploy match {
+                   case Nil =>
+                     Sync[F].raiseError[(List[(DeployId, BlockHash)], Crc32[F])](
+                       BlockHashesByDeployLogIsCorrupted
+                     )
+                   case _ :: _ =>
+                     // Trying to delete the last log entry
+                     for {
```

Comment:
is that method deleting anything? If so, why is it called 'validate'?

Or maybe we allow some inconsistency for the last log entry? If so, why?

Reply:
The node could have crashed while writing the last log entry. If it did, then the node didn't update the CRC value located in a separate file. Which means we can delete the last entry and check if it matches the written CRC value.

Reply:
what happens if I fail here, and then restart the node again?

Also, this looks like a generic concern for the indexing store - all logs should be the same in that matter, and it should be handled by the same code :)

Reply:
Nothing will really change. The last entry garbage which was there previously will potentially be overwritten by some new garbage. After restarting the node will realise that the CRC does not match and will try to delete the last entry once again.